docs: Explore test-targets-for-shards SDD (#1627)
Flank/flank: docs/feature/1532-explore-test-target-for-shards-with-extra-features.md

# Auto discover rest of tests when using test-target-for-shards option

Currently, when using `test-target-for-shards` you need to specify all tests manually to split to shards. If you do not
do that then only listed tests are executed.
It will be good to provide an option to auto-discover other than specified tests to make `test-target-for-shards`
better and easier to use.

# References

- [Conversation on Slack](https://firebase-community.slack.com/archives/C72V6UW8M/p1611322014000600)
- [1532](https://github.com/Flank/flank/issues/1532)

# Motivation

`test-target-for-shards` is not so easy to use, this change will improve it when using Flank and give users more
flexibility and control over creating shards.

# Goals

1. New option is available on the flank configuration level
1. User could specify which packages/class/tests run on specific shards by using a new option

> 1. Other not specified tests are automatically added to the left shard/shards

Reviewer: ```suggestion
1. Other tests that are not specified are then automatically added to a different set of shards/shard
```

# Design

> - Add new option under Flank configuration, possible ideas are: `plan-sharding`, `tests-for-shards`.

Reviewer: ```suggestion
- Add new option under Flank configuration, possible ideas are: `plan-sharding`, `tests-for-shards`, `smart-test-targets-for-shards`.
```

> - New option should behave the same as [`test-target-for-shards`](https://cloud.google.com/sdk/gcloud/reference/beta/firebase/test/android/run#--test-targets-for-shard).

Reviewer: ```suggestion
- New option should at least behave the same as [`test-target-for-shards`](https://cloud.google.com/sdk/gcloud/reference/beta/firebase/test/android/run#--test-targets-for-shard).
```

> - After that Flank should split the test into shards specified by the user and make another shard/shards with the left one.

Reviewer: ```suggestion
- The left over tests should then be sorted into other shards or a single shard based upon the configuration provided by the user.
```

- Flank should fast fail if `disable-sharding` is set to `true`  or `max-test-shards` is lower than specified test shards
  by user + 1.

# API

> It is hard to plan API for that change because the codebase could be different when doing this task and also usage of this

Reviewer: ```suggestion
It is hard to plan for this change to the API because by the time this is worked upon the codebase may have changed significantly influencing the usage of this request.
```

> the option will be applied to many places in code.

Reviewer: ```suggestion
This option will be applied to many places in code.
```

Please follow up `Design` section for the implementation plan.

# Results

> A new option which split sharding based on user input or alternative flag (described in `Alternative Considered`) which

Reviewer: ```suggestion
A new option will be created which will split sharding based on user input or an alternative flag (described in `Alternative Considered`) which
```

will change behavior of `test-target-for-shards`.

# Dependencies

> There are not any dependencies that could have a real impact on this task. However, it will be best to do this after

Reviewer: ```suggestion
There are no dependencies that will have a impact on this task. It is recommended to do this after 
```
